src/server/main_service.cc

```diff
@@ -1147,6 +1154,19 @@ void Service::DispatchCommand(CmdArgList args, facade::ConnectionContext* cntx)
     dfly_cntx->reply_builder()->CloseConnection();
   }
 
+  // if this is a read command, and client tracking has enabled,
+  // start tracking all the updates to the keys in this read command
+  if ((cid->opt_mask() & CO::READONLY) && dfly_cntx->conn()->IsTrackingOn()) {
+    auto cb = [&](Transaction* t, EngineShard* shard) {
+      auto keys = t->GetShardArgs(shard->shard_id());
+      vector<string_view> keys_to_track{keys.begin(), keys.end()};
+      return OpTrackKeys(t->GetOpArgs(shard), dfly_cntx, keys_to_track);
+    };
+
+    dfly_cntx->transaction->Refurbish();
```

Comment:
1. Refurbish() should be after you use it with schedulesinglehop, please add a test with move, zinterstore etc (commands that have two hops)
2. DispatchCommand is not used for all commands, but we can ignore that for now
3. Still reminding that we can track keys when the transaction concludes and no "hop" is needed solely for this, but I also assume it's not a must for the first version

Reply:
> 1. Refurbish() should be after you use it with schedulesinglehop, please add a test with move, zinterstore etc (commands that have two hops)

I'm wondering, the transaction probably has already been used before the tracking (e.g. in the previous command invoking?) and I need to do a refurbish here before doing my own singlehop?

Reply:
> 1. Refurbish() should be after you use it with schedulesinglehop, please add a test with move, zinterstore etc (commands that have two hops)
> 2. DispatchCommand is not used for all commands, but we can ignore that for now
> 3. Still reminding that we can track keys when the transaction concludes and no "hop" is needed solely for this, but I also assume it's not a must for the first version

Sounds good to me, I will look into this later.

Reply:
> 1. please add a test with move, zinterstore etc (commands that have two hops)

sounds good, I will include all the functional tests in the second PR, which will completes the implementation with invalidation logic.